## This project detects spam messages using vectorizing of text on training data and test using logistic regression model

In [10]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report
from sklearn.feature_extraction.text import TfidfVectorizer

1.Dataset Overview..

In [11]:
df=pd.read_csv('spamdata.csv',encoding='latin-1')
df = df[['v1', 'v2']]
df.columns = ['label', 'message']
X = df['message']
y = df['label']

print(df.shape)
print(df['label'].value_counts())  # how many spam vs ham
df.head()

X = df['message']
y = df['label'] 
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

(5572, 2)
label
ham     4825
spam     747
Name: count, dtype: int64


2. Preprocessing..
    we use TF-IDF to vectorize text messages

In [ ]:
tfidf=TfidfVectorizer()
X_train_tfidf=tfidf.fit_transform(X_train)
X_test_tfidf=tfidf.transform(X_test)
print(X_train_tfidf.shape)

(4457, 7735)


3.Model Training.. and testing

In [ ]:
model = LogisticRegression()
model.fit(X_train_tfidf,y_train)

predictions = model.predict(X_test_tfidf)
print(accuracy_score(y_test, predictions))
print(classification_report(y_test, predictions))

0.967713004484305
              precision    recall  f1-score   support

         ham       0.96      1.00      0.98       965
        spam       0.99      0.77      0.86       150

    accuracy                           0.97      1115
   macro avg       0.98      0.88      0.92      1115
weighted avg       0.97      0.97      0.97      1115



4.Comparison with Naive Bayes

In [14]:
from sklearn.naive_bayes import MultinomialNB
nb_model = MultinomialNB()
nb_model.fit(X_train_tfidf, y_train)
nb_predictions = nb_model.predict(X_test_tfidf)
print("Naive Bayes:", accuracy_score(y_test, nb_predictions))
print("Logistic Regression:", accuracy_score(y_test, predictions))

Naive Bayes: 0.9623318385650225
Logistic Regression: 0.967713004484305


In [15]:
messages = [
    "Win a free iPhone click here now",           # obvious spam
    "hey are you coming to class tomorrow",        # obvious ham  
    "Your account needs verification urgently",    # tricky spam
    "Meeting rescheduled to 3pm let me know"       # tricky ham
]

for msg in messages:
    vec = tfidf.transform([msg])
    pred = model.predict(vec)
    print(f"{msg[:40]}... → {pred[0]}")

Win a free iPhone click here now... → spam
hey are you coming to class tomorrow... → ham
Your account needs verification urgently... → ham
Meeting rescheduled to 3pm let me know... → ham


## Model Limitations
- Model misclassifies sophisticated phishing messages 
  like "Your account needs verification urgently"
- Trained on SMS data so may not generalize to email spam
- TF-IDF doesn't understand context, only word frequency
- A deep learning model like BERT would handle these better

## Conclusion
- Logistic Regression (96.7%) slightly outperformed Naive Bayes (96.2%)
- Precision matters more than recall for spam detection to avoid losing important messages
- TF-IDF + Logistic Regression is a strong baseline for text classification
- Future improvement: fine-tune a BERT model to catch sophisticated phishing messages